File: CF/G.py
import math

freeId = 1

is_gini = False

class Node:
    featureNumber = 0
    b = -100_000
    classAns = -1
    is_c = False

    leftChild = None
    rightChild = None


    # list of teach
    # hightOst of visota do 0
    def __init__(self, teach, hightOst, k):
        global freeId
        self.id = freeId
        freeId += 1
        self.k = k
        self.hightOst = hightOst
        self.buildNode(teach, hightOst)

    def initClassesMap(self, k):
        return {i: 0 for i in range(k)}

    def findMaxClassByChastota(self, teach):
        tmp = self.initClassesMap(self.k)
        for el in teach:
            tmp[el[1]] += 1
        bestKey = 0
        for key in tmp:
            if tmp[key] > tmp[bestKey]:
                bestKey = key
        return bestKey

    def buildNode(self, teach, hightOst):
        if hightOst == 1:
            self.classAns = self.findMaxClassByChastota(teach)
            return
        bestFeature = 0
        bestB = 100_000
        bestScore = 100_000

        all_indexes = self.initClassesMap(self.k)
        for i in range(len(teach)):
            all_indexes[teach[i][1]] += 1

        left_best_ind = 0

        for featureI in range(len(teach[0][0])):
            hello = []
            for i in range(len(teach)):
                hello.append((teach[i][0][featureI], i))
            hello = sorted(hello)
            left_ind = 0
            leftClassesMap = self.initClassesMap(self.k)
            left_classes_count_con = 0
            right_classes_count_con = len(hello)


            while left_ind < len(hello):
                znash = hello[left_ind][0]
                while left_ind < len(hello) and znash == hello[left_ind][0]:
                    left_classes_count_con += 1
                    right_classes_count_con -= 1
                    leftClassesMap[teach[hello[left_ind][1]][1]] += 1
                    left_ind += 1

                if is_gini:
                    score = self.evalGini(leftClassesMap, left_classes_count_con, all_indexes, is_left=True) * left_classes_count_con + self.evalGini(leftClassesMap, right_classes_count_con, all_indexes, is_left=False) * right_classes_count_con
                else:
                    score = self.evalEntrop(leftClassesMap, left_classes_count_con, all_indexes, is_left=True) * left_classes_count_con + self.evalEntrop(leftClassesMap, right_classes_count_con, all_indexes, is_left=False) * right_classes_count_con
                # print(f'{score} {featureI}')
                if bestScore > score:
                    bestScore = score
                    left_best_ind = left_ind
                    bestB = hello[left_ind][0]
                    bestFeature = featureI

        hello = []
        for i in range(len(teach)):
            hello.append((teach[i][0][bestFeature], i))
        hello = sorted(hello)
        print(f'bf {bestFeature}')

        leftListBEst = list(map(lambda e: teach[e[1]], hello[:left_best_ind]))
        rightListBEst = list(map(lambda e: teach[e[1]], hello[left_best_ind:]))

        if len(leftListBEst) == 0 or len(rightListBEst) == 0:
            self.is_c = True
            return
        # print(f'l {len(leftListBEst)}')
        # print(f'r {len(rightListBEst)}')
        leftNode = Node(leftListBEst, hightOst - 1, self.k)
        rightNode = Node(rightListBEst, hightOst - 1, self.k)
        self.b = bestB
        self.featureNumber = bestFeature
        self.leftChild = leftNode
        self.rightChild = rightNode

    # list size of k different classes and count objects
    def evalEntrop(self, childs, elemsSum, all_indexes, is_left):
        res = 0
        for key in childs:
            if is_left:
                x = childs[key]
            else:
                x = all_indexes[key] - childs[key]
            if x != 0:
                p = x * 1.0 / elemsSum
                res -= p * math.log(p)
        return res

    def evalGini(self, childs, elemsSum, all_indexes, is_left):
        res = 0
        for key in childs:
            if is_left:
                x = childs[key]
            else:
                x = all_indexes[key] - childs[key]
            if x != 0:
                p = x * 1.0 / elemsSum
                res += p ** 2
        return 1 - res

    def printNode(self, listRes):
        if self.hightOst == 1 or self.is_c == True:
            listRes.append((self.id, f'C {self.classAns + 1}'))
        else:
            listRes.append((self.id, f'Q {self.featureNumber + 1} {self.b} {self.leftChild.id} {self.rightChild.id}'))
            self.leftChild.printNode(listRes)
            self.rightChild.printNode(listRes)


class Tree:
    head = None

    def __init__(self, teach, maxH, k):
        self.buildTree(teach, maxH + 1, k)

    def buildTree(self, teach, maxH, k):
        self.head = Node(teach, maxH, k)

    def printTree(self):
        listRes = list()
        self.head.printNode(listRes)
        listRes = sorted(listRes, key=lambda el: el[0])

        print(freeId - 1)
        for el in listRes:
            print(el[1])


if __name__ == '__main__':
    m, k, h = map(int, input().split(" "))
    n = int(input())
    teach = list()
    for i in range(n):
        inp = list(map(int, input().split(" ")))
        features = inp[:m]
        clazz = inp[m]
        teach.append((features, clazz - 1))
    if len(teach) > 200:
        is_gini = True
    treeML = Tree(teach, h, k)
    treeML.printTree()

    # print(teach)

'''
Answer:

7
Q 1 2.5 2 5
Q 2 2.5 3 4
C 1
C 4
Q 2 2.5 6 7
C 2
C 3
'''

File: CF/G.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <map>
#include <cmath>

using namespace std;

int m, k, h, n, tmp1;
int freeId = 1;

bool is_gini = false;
vector<pair<vector<int>, int>> teach;


struct Node {
    int id = 0;
    int hightOst = 0;

    int featureNumber = 0;
    int b = -100000;
    int classAns = -1;

    bool is_c = false;

    Node *leftChild = nullptr;
    Node *rightChild = nullptr;


    Node(const vector<int> &indexes, int hightOstH) {
        id = freeId;
        hightOst = hightOstH;
        ++freeId;
        buildNode(indexes);
    }

    static unordered_map<int, int> initClassesMap() {
        unordered_map<int, int> maps;
        for (size_t i = 0; i < k; i++) {
            maps[i] = 0;
        }
        return maps;
    }

    static int findMaxClassByChastota(const vector<int> &indexes) {
        auto tmp = initClassesMap();
        for (auto &el: indexes) {
            tmp[teach[el].second] += 1;
        }
        int bestKey = 0;
        for (auto &p : tmp) {
            if (p.second > tmp[bestKey]) {
                bestKey = p.first;
            }
        }
        return bestKey;
    }

    void buildNode(const vector<int> &indexes) {
        if (hightOst == 1) {
            classAns = findMaxClassByChastota(indexes);
            return;
        }
        int bestFeature = 0;
        int bestB = 100000;
        double bestScore = 100000;

        auto all_indexes = initClassesMap();

        for (auto &el: indexes) {
            all_indexes[teach[el].second] += 1;
        }

        int left_best_ind = 0;

        for (size_t featureI = 0; featureI < m; featureI++) {
            vector<pair<int, int>> hello;
            hello.reserve(indexes.size());

            for (auto &el: indexes) {
                hello.emplace_back(teach[el].first[featureI], el);
            }
            sort(hello.begin(), hello.end());
            int left_ind = 0;

            auto leftClassesMap = initClassesMap();
            int left_classes_count_con = 0;
            int right_classes_count_con = hello.size();
            double score;

            while (left_ind < hello.size()) {
                int znash = hello[left_ind].first;
                while (left_ind < hello.size() && znash == hello[left_ind].first) {
                    left_classes_count_con += 1;
                    right_classes_count_con -= 1;
                    leftClassesMap[teach[hello[left_ind].second].second] += 1;
                    left_ind += 1;
                }

                if (is_gini) {
                    score = evalGini(leftClassesMap, left_classes_count_con, all_indexes, true) * left_classes_count_con +
                            evalGini(leftClassesMap, right_classes_count_con, all_indexes, false) * right_classes_count_con;
                } else {
                    score = evalEntrop(leftClassesMap, left_classes_count_con, all_indexes, true) *
                            left_classes_count_con +
                            evalEntrop(leftClassesMap, right_classes_count_con, all_indexes, false) *
                            right_classes_count_con;
                }
                if (bestScore > score) {
                    bestScore = score;
                    left_best_ind = left_ind;
                    bestB = hello[left_ind].first;
                    bestFeature = featureI;
                }
            }
        }

        vector<pair<int, int>> hello;
        hello.reserve(indexes.size());
        for (auto &el: indexes) {
            hello.emplace_back(teach[el].first[bestFeature], el);
        }
        sort(hello.begin(), hello.end());

        vector<int> leftListBEst, rightListBEst;
        for (size_t i = 0; i < hello.size(); i++) {
            if (i < left_best_ind) {
                leftListBEst.push_back(hello[i].second);
            } else {
                rightListBEst.push_back(hello[i].second);
            }
        }

        if (leftListBEst.empty() || rightListBEst.empty()) {
            classAns = findMaxClassByChastota(indexes);
            is_c = true;
            return;
        }
        leftChild = new Node(leftListBEst, hightOst - 1);
        rightChild = new Node(rightListBEst, hightOst - 1);
        b = bestB;
        featureNumber = bestFeature;
    }


    static double
    evalEntrop(unordered_map<int, int> &childs, int elemsSum, unordered_map<int, int> &all_indexes, bool is_left) {
        double res = 0;
        double p;
        int x;
        for (auto &el: childs) {
            if (is_left) {
                x = el.second;
            } else {
                x = all_indexes[el.first] - el.second;
            }
            if (x != 0) {
                p = static_cast<double>(x) / elemsSum;
                res -= (p * log(p));
            }
        }
        return res;
    }

    static double
    evalGini(unordered_map<int, int> &childs, int elemsSum, unordered_map<int, int> &all_indexes, bool is_left) {
        double res = 0;
        double p;
        int x;
        for (auto &el: childs) {
            if (is_left) {
                x = el.second;
            } else {
                x = all_indexes[el.first] - el.second;
            }
            if (x != 0) {
                p = static_cast<double>(x) / elemsSum;
                res += (p * p);
            }
        }
        return 1 - res;
    }

    void printNode(vector<pair<int, string>>& listRes) {
        if (hightOst == 1 || is_c) {
            listRes.emplace_back(id, "C " + to_string(classAns + 1));
        } else {
            listRes.emplace_back(id, "Q " + to_string(featureNumber + 1) + " " + to_string(b) + " " + to_string(leftChild->id)+ " " + to_string(rightChild->id));
            leftChild->printNode(listRes);
            rightChild->printNode(listRes);
        }
    }
};

struct Tree {
    Node *head = nullptr;

    Tree() {
        vector<int> indexes;
        for(size_t i = 0; i < teach.size(); i++) {
            indexes.push_back(i);
        }
        head = new Node(indexes, h + 1);
    }


    void printTree() const {
        vector<pair<int, string>> results;
        head->printNode(results);
        sort(results.begin(),results.end());

        cout << (freeId - 1) << endl;
        for(auto& el: results) {
            cout << el.second << endl;
        }
    }
};


int main() {
    cin >> m >> k >> h >> n;
    teach.reserve(n);
    for (size_t i = 0; i < n; i++) {
        vector<int> features;
        features.reserve(m);
        for (size_t j = 0; j < m; j++) {
            cin >> tmp1;
            features.emplace_back(tmp1);
        }
        cin >> tmp1;
        teach.emplace_back(features, tmp1 - 1);
    }

    if (teach.size() > 50) {
        is_gini = true;
    }
    Tree tree_ML = Tree();
    tree_ML.printTree();

    return 0;
}